README

See http://chunkfs.florz.de/ for more information and new versions.



debian/changelog
chunkfs (0.8) unstable; urgency=low

  * Build-depend on perl for any arch to support cross building.
  * Adapt writeoverlay.sh to changed losetup options.
  * Use portable format specifiers for big integers.
  * Update to debhelper compatibility level 9.

 -- Florian Zumbiehl <florz@florz.de>  Thu, 13 Jun 2019 12:35:35 +0200

chunkfs (0.7) unstable; urgency=low

  * Check that file descriptors 0 to 2 are open on startup.
  * Change Debian dependency from deprecated fuse-utils to fuse.

 -- Florian Zumbiehl <florz@florz.de>  Wed, 25 Dec 2013 17:43:35 +0100

chunkfs (0.6) unstable; urgency=low

  * Fix potential filedescriptor leak in unchunkfs_read() in case of
    pread() errors.

 -- Florian Zumbiehl <florz@florz.de>  Wed, 17 Mar 2010 21:02:55 +0100

chunkfs (0.5) unstable; urgency=low

  * Mostly style improvements, in particular more useful help output
  * writeoverlay.sh now uses new losetup option -s to avoid race
    condition in device allocation

 -- Florian Zumbiehl <florz@florz.de>  Sat, 29 Aug 2009 02:11:08 +0200

chunkfs (0.4) unstable; urgency=low

  * fixed read errors being ignored because of size_t being used where
    ssize_t should be

 -- Florian Zumbiehl <florz@florz.de>  Wed, 23 Jul 2008 10:18:36 +0200

chunkfs (0.3) unstable; urgency=low

  * Added writeoverlay.sh

 -- Florian Zumbiehl <florz@florz.de>  Thu,  5 Jul 2007 03:48:03 +0200

chunkfs (0.2) unstable; urgency=low

  * Added UnChunkFS

 -- Florian Zumbiehl <florz@florz.de>  Wed, 27 Jun 2007 00:08:38 +0200

chunkfs (0.1) unstable; urgency=low

  * Initial release

 -- Florian Zumbiehl <florz@florz.de>  Mon, 20 Jun 2007 18:29:09 +0200



chunkfs.c
/*
 *  ChunkFS - mount arbitrary files via FUSE as a tree of chunk files
 *  Copyright (C) 2007-2013  Florian Zumbiehl <florz@florz.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#define _FILE_OFFSET_BITS 64
#define _XOPEN_SOURCE 500
#define FUSE_USE_VERSION 25

#include <sys/types.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <stdio.h>
#include <limits.h>
#include <stdint.h>
#include <inttypes.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <syslog.h>
#include <stdbool.h>
#include <sys/ioctl.h>
#include <linux/fs.h>

#include <fuse.h>

#include "utils.h"

struct chunk_stat {
	int level;
	mode_t mode;
	off_t chunk,offset,size;
};

static int image_fd;
static off_t chunk_size,image_size,image_chunks;

static int resolve_path(const char *path,struct chunk_stat *st){
	size_t l;
	uint64_t r;
	uint8_t tmp;

	l=strlen(path);
	r=0;
	if(l<3||l>8*3||l%3){
		if(strcmp(path,"/"))return -ENOENT;
	}else{
		for(int x=0;x<l;x++){
			tmp=path[x];
			if(x%3){
				if(tmp>='0'&&tmp<='9'){
					tmp-='0';
				}else if(tmp>='a'&&tmp<='f'){
					tmp-='a'-10;
				}else{
					return -ENOENT;
				}
				r|=((uint64_t)tmp)<<((8*2-x/3*2-x%3)*4);
			}else if(tmp!='/')return -ENOENT;
		}
		if(r>=(uint64_t)image_chunks)return -ENOENT;
	}
	st->level=l/3;
	st->chunk=r;
	st->offset=st->chunk*chunk_size;
	if(st->level<8){
		st->mode=S_IFDIR;
	}else{
		st->mode=S_IFREG;
		st->size=min(image_size-st->offset,chunk_size);
	}
	return 0;
}

static int chunkfs_getattr(const char *path,struct stat *buf){
	int r;
	struct chunk_stat st;

	if((r=resolve_path(path,&st))<0)return r;
	if(fstat(image_fd,buf)<0)return -errno;
	buf->st_mode=(buf->st_mode&~(S_IFMT|S_IXUSR|S_IXGRP|S_IXOTH))|(st.mode&S_IFMT);
	if(S_ISDIR(st.mode)){
		buf->st_mode=buf->st_mode
			|((buf->st_mode&S_IRUSR)?S_IXUSR:0)
			|((buf->st_mode&S_IRGRP)?S_IXGRP:0)
			|((buf->st_mode&S_IROTH)?S_IXOTH:0);
		buf->st_nlink=256+2;
		buf->st_size=0;
	}else{
		buf->st_nlink=1;
		buf->st_size=st.size;
	}
	buf->st_blocks=0;
	return 0;
}

static int chunkfs_readdir(const char *path,void *buf,fuse_fill_dir_t filler,off_t offset,struct fuse_file_info *fi){
	int r;
	struct chunk_stat st;
	uint64_t chunks_per_entry;

	if((r=resolve_path(path,&st))<0)return r;
	if(!S_ISDIR(st.mode))return -ENOTDIR;
	chunks_per_entry=1ULL<<((8-1-st.level)*8);
	filler(buf,".",NULL,0);
	filler(buf,"..",NULL,0);
	for(uint64_t x=0;x<256&&(uint64_t)st.chunk+x*chunks_per_entry<(uint64_t)image_chunks;x++){
		char nbuf[3];
		sprintf(nbuf,"%02" PRIx64,x);
		filler(buf,nbuf,NULL,0);
	}
	return 0;
}

static int chunkfs_open(const char *path,struct fuse_file_info *fi){
	int r;
	struct chunk_stat st;

	if((r=resolve_path(path,&st))<0)return r;
	if((fi->flags&O_ACCMODE)!=O_RDONLY)return -EROFS;
	return 0;
}

static int chunkfs_read(const char *path,char *buf,size_t count,off_t offset,struct fuse_file_info *fi){
	int r;
	struct chunk_stat st;
	ssize_t bufpos,r2;

	if((r=resolve_path(path,&st))<0)return r;
	if(S_ISDIR(st.mode))return -EISDIR;
	if(count>INT_MAX)die(false,"read request larger than can be represented in an int");
	if(count>SSIZE_MAX)die(false,"read request larger than can be represented in an ssize_t");
	count=min((off_t)count,max(st.size-offset,(off_t)0));
	for(bufpos=0;bufpos<count;){
		if((r2=pread(image_fd,buf+bufpos,count-bufpos,st.offset+offset+(off_t)bufpos))<0)return -errno;
		if(!r2)return -EIO;
		bufpos+=r2;
	}
	return bufpos;
}

static struct fuse_operations chunkfs_ops={
	.getattr=chunkfs_getattr,
	.readdir=chunkfs_readdir,
	.open=chunkfs_open,
	.read=chunkfs_read
};

int main(int argc,char **argv){
	bool validation_error=false,show_version=false,show_help=false;
	int opt,ret=0;
	char *fake_argv[2]={"",NULL};
	char *end;
	struct stat st;

	{
		int fd;

		if((fd=dup(0))<3||close(fd)<0)_exit(1);
	}

	openlog("chunkfs",LOG_CONS|LOG_NDELAY|LOG_PERROR|LOG_PID,LOG_DAEMON);
	while((opt=getopt(argc,argv,":hVdfso:"))!=-1)
		switch(opt){
			case 'h':
				show_help=true;
				break;
			case 'V':
				show_version=true;
				break;
			case '?':
				fprintf(stderr,"chunkfs: invalid option: -%c\n",optopt);
				validation_error=true;
				break;
			case ':':
				fprintf(stderr,"chunkfs: option requires an argument: -%c\n",optopt);
				validation_error=true;
				break;
		}
	if(show_help+show_version>1||argc-optind!=(show_help||show_version?0:3))validation_error=true;
	if(validation_error||show_help){
		fprintf(stderr,
			"Usage: %s [options] <chunk size> <image file> <mount point>\n"
			"\n"
			"general options:\n"
			"    -o opt[,opt...]        mount options\n"
			"    -h                     print help\n"
			"    -V                     print version\n"
			"\n",
			*argv);
		fake_argv[1]="-ho";
		fuse_main(2,fake_argv,&chunkfs_ops);
		ret=validation_error;
	}else if(show_version){
		fprintf(stderr,"ChunkFS v" VERSION "\n");
		fake_argv[1]="-V";
		fuse_main(2,fake_argv,&chunkfs_ops);
	}else{
		errno=0;
		chunk_size=strtoll(argv[optind],&end,10);
		if(errno||*end||chunk_size<1)die(false,"Specified invalid chunk size");
		if((image_fd=open(argv[optind+1],O_RDONLY))<0)die(true,"open(image)");
		if(fstat(image_fd,&st)<0)die(true,"stat(image)");
		if(S_ISBLK(st.st_mode)){
			uint64_t blksize;
			if(ioctl(image_fd,BLKGETSIZE64,&blksize)<0)die(true,"ioctl(image,BLKGETSIZE64)");
			if(blksize>(uint64_t)INT64_MAX)die(false,"block device too large");
			image_size=blksize;
		}else{
			image_size=st.st_size;
		}
		image_chunks=image_size/chunk_size;
		if(image_size%chunk_size)image_chunks++;
		argv[optind]=argv[optind+2];
		argc-=2;
		ret=fuse_main(argc,argv,&chunkfs_ops);
	}
	return ret;
}



writeoverlay.sh
#!/bin/sh
#
# Copyright (C) 2007-2010  Florian Zumbiehl <florz@florz.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

#
# !!! Be warned that this script isn't very sophisticated in terms of     !!!
# !!! error handling and the handling of special characters in arguments! !!!
#
# This script allows you to create from an arbitrary block device a
# writable clone by the means of the Linux device mapper's snapshot
# facilities. Any writes to the clone will be stored in a temporary file.
#
# The primary intended use for this is to allow for read-only filesystem
# images (like from backups) to be mounted, even if they need some form
# of recovery, like a log replay in case of a journalling filesystem.
# That's a common problem if the image was created from a block device
# snapshot of a live filesystem.
#
# If the image is a regular file, you'll first have to create a loop
# device from that, and then use this script on that loop device.
#

set -e
#set -x

usage(){
	echo usage: $0 '[ create <origdev> <cowsize> <cow_sect_per_chunk> <mapper_dev_name> | remove <mapper_dev_name> ]'
	exit 1
}

case "$1" in
	create)
		[ $# -eq 5 ] || usage
		ORIGDEV=$2
		COWSIZE=$3
		SECT_PER_CHUNK=$4
		OVERLAYDEV=$5
		COWFILE=`mktemp -t writeoverlay.XXXXXXXXXX`
		dd if=/dev/zero bs=1 seek=`echo $COWSIZE-1 | bc` count=1 of=$COWFILE > /dev/null
		LOOPDEV=`losetup -f --show $COWFILE`
		rm $COWFILE
		echo "0 `blockdev --getsz $ORIGDEV` snapshot $ORIGDEV $LOOPDEV n $SECT_PER_CHUNK" | dmsetup create $OVERLAYDEV
		;;
	remove)
		[ $# -eq 2 ] || usage
		OVERLAYDEV=$2
		LOOPDEV=/dev/loop`dmsetup table $OVERLAYDEV | cut '-d ' -f5 | cut -d: -f2`
		dmsetup remove $OVERLAYDEV
		losetup -d $LOOPDEV
		;;
	*)
		usage
		;;
esac



unchunkfs.c
/*
 *  ChunkFS - mount arbitrary files via FUSE as a tree of chunk files
 *  Copyright (C) 2007-2013  Florian Zumbiehl <florz@florz.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#define _FILE_OFFSET_BITS 64
#define _XOPEN_SOURCE 500
#define FUSE_USE_VERSION 25

#include <sys/types.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <stdio.h>
#include <limits.h>
#include <stdint.h>
#include <inttypes.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <syslog.h>
#include <stdbool.h>
#include <sys/ioctl.h>
#include <linux/fs.h>

#include <fuse.h>

#include "utils.h"

static int chunkdir_fd;
static off_t chunk_size,image_size;

static int ensure_chunkdir_is_cwd(void){
	static bool done;

	if(!done){
		if(fchdir(chunkdir_fd)<0)return -errno;
		done=true;
	}
	return 0;
}

static int resolve_path(const char *path,mode_t *mode){
	if(!strcmp(path,"/")){
		*mode=S_IFDIR;
	}else if(!strcmp(path,"/image")){
		*mode=S_IFREG;
	}else{
		return -ENOENT;
	}
	return 0;
}

static void gen_chunk_name(char *buf,off_t num){
	sprintf(buf+7,"%016" PRIxMAX,(intmax_t)num);
	for(int x=0;x<7;x++){
		memmove(buf+x*3,buf+7+x*2,2);
		buf[x*3+2]='/';
	}
}

static int unchunkfs_getattr(const char *path,struct stat *buf){
	int r;
	mode_t mode;

	if((r=resolve_path(path,&mode))<0)return r;
	if(fstat(chunkdir_fd,buf)<0)return -errno;
	buf->st_mode=(buf->st_mode&~(S_IFMT|S_IXUSR|S_IXGRP|S_IXOTH))|(mode&S_IFMT);
	if(S_ISDIR(mode)){
		buf->st_mode=buf->st_mode
			|((buf->st_mode&S_IRUSR)?S_IXUSR:0)
			|((buf->st_mode&S_IRGRP)?S_IXGRP:0)
			|((buf->st_mode&S_IROTH)?S_IXOTH:0);
		buf->st_nlink=2;
		buf->st_size=0;
	}else{
		buf->st_nlink=1;
		buf->st_size=image_size;
	}
	buf->st_blocks=0;
	return 0;
}

static int unchunkfs_readdir(const char *path,void *buf,fuse_fill_dir_t filler,off_t offset,struct fuse_file_info *fi){
	int r;
	mode_t mode;

	if((r=resolve_path(path,&mode))<0)return r;
	if(!S_ISDIR(mode))return -ENOTDIR;
	filler(buf,".",NULL,0);
	filler(buf,"..",NULL,0);
	filler(buf,"image",NULL,0);
	return 0;
}

static int unchunkfs_open(const char *path,struct fuse_file_info *fi){
	int r;
	mode_t mode;

	if((r=resolve_path(path,&mode))<0)return r;
	if((fi->flags&O_ACCMODE)!=O_RDONLY)return -EROFS;
	return 0;
}

static int unchunkfs_read(const char *path,char *buf,size_t count,off_t offset,struct fuse_file_info *fi){
	int r,chunk_fd;
	mode_t mode;
	off_t chunk_num,chunk_offset;
	char chunk_name[24];
	ssize_t bufpos,r2;

	if(ensure_chunkdir_is_cwd()<0)return -EIO;
	if((r=resolve_path(path,&mode))<0)return r;
	if(S_ISDIR(mode))return -EISDIR;
	if(count>INT_MAX)die(false,"read request larger than can be represented in an int");
	if(count>SSIZE_MAX)die(false,"read request larger than can be represented in an ssize_t");
	count=min((off_t)count,max(image_size-offset,(off_t)0));
	for(bufpos=0;bufpos<count;){
		chunk_num=(offset+(off_t)bufpos)/chunk_size;
		chunk_offset=(offset+(off_t)bufpos)%chunk_size;
		gen_chunk_name(chunk_name,chunk_num);
		if((chunk_fd=open(chunk_name,O_RDONLY))<0)return -EIO;
		if((r2=pread(chunk_fd,buf+bufpos,count-bufpos,chunk_offset))>0)bufpos+=r2;
		if(close(chunk_fd)<0)die(true,"close(chunk_fd)");
		if(r2<=0)return -(r2?errno:EIO);
	}
	return bufpos;
}

static struct fuse_operations unchunkfs_ops={
	.getattr=unchunkfs_getattr,
	.readdir=unchunkfs_readdir,
	.open=unchunkfs_open,
	.read=unchunkfs_read
};

int main(int argc,char **argv){
	bool validation_error=false,show_version=false,show_help=false;
	int opt,ret=0;
	char *fake_argv[2]={"",NULL};
	int cwd_fd;
	struct stat st;
	off_t last_chunk,last_chunk_size;

	{
		int fd;

		if((fd=dup(0))<3||close(fd)<0)_exit(1);
	}

	openlog("unchunkfs",LOG_CONS|LOG_NDELAY|LOG_PERROR|LOG_PID,LOG_DAEMON);
	while((opt=getopt(argc,argv,":hVdfso:"))!=-1)
		switch(opt){
			case 'h':
				show_help=true;
				break;
			case 'V':
				show_version=true;
				break;
			case '?':
				fprintf(stderr,"unchunkfs: invalid option: -%c\n",optopt);
				validation_error=true;
				break;
			case ':':
				fprintf(stderr,"unchunkfs: option requires an argument: -%c\n",optopt);
				validation_error=true;
				break;
		}
	if(show_help+show_version>1||argc-optind!=(show_help||show_version?0:2))validation_error=true;
	if(validation_error||show_help){
		fprintf(stderr,
			"Usage: %s [options] <chunk dir> <mount point>\n"
			"\n"
			"general options:\n"
			"    -o opt[,opt...]        mount options\n"
			"    -h                     print help\n"
			"    -V                     print version\n"
			"\n",
			*argv);
		fake_argv[1]="-ho";
		fuse_main(2,fake_argv,&unchunkfs_ops);
		ret=validation_error;
	}else if(show_version){
		fprintf(stderr,"UnChunkFS v" VERSION "\n");
		fake_argv[1]="-V";
		fuse_main(2,fake_argv,&unchunkfs_ops);
	}else{
		if((cwd_fd=open(".",O_RDONLY))<0)die(true,"open(.)");
		if((chunkdir_fd=open(argv[optind],O_RDONLY))<0)die(true,"open(<chunk dir>)");
		if(fchdir(chunkdir_fd)<0)die(true,"fchdir(<chunk dir>)");
		last_chunk=0;
		last_chunk_size=0;
		for(int x=64;x--;){
			off_t new_last_chunk;
			char chunk_name[24];

			new_last_chunk=last_chunk|(x<63?(off_t)1<<x:0);
			gen_chunk_name(chunk_name,new_last_chunk);
			if(stat(chunk_name,&st)<0){
				if(errno!=ENOENT)die(true,"stat(<chunk dir>/%s)",chunk_name);
			}else{
				if(st.st_size<1)die(false,"<chunk dir>/%s is smaller than one byte",chunk_name);
				if(!chunk_size)chunk_size=st.st_size;
				last_chunk=new_last_chunk;
				last_chunk_size=st.st_size;
			}
		}
		if(chunk_size<1)chunk_size=1;
		if(last_chunk>ANYSINT_MAX(off_t)/chunk_size)die(false,"the sum of the chunks is too large");
		image_size=last_chunk*chunk_size;
		if(last_chunk_size>ANYSINT_MAX(off_t)-image_size)die(false,"the sum of the chunks is too large");
		image_size+=last_chunk_size;
		argv[optind]=argv[optind+1];
		argc--;
		if(fchdir(cwd_fd)<0)die(true,"fchdir(<start cwd>)");
		ret=fuse_main(argc,argv,&unchunkfs_ops);
	}
	return ret;
}

